rfc/rfc-61/rfc-61.md

```diff
@@ -0,0 +1,98 @@
+that means, multiple writers prune to write into the same **small** buckets at the same time, that's how conflicts happen.
+
+In general, for multiple streaming writers ingestion, explicit lock is not very capable of putting into production, in this RFC, we propse a lockless solution for streaming ingestion.
+
+## Background
+
+Streaming jobs are naturally suitable for data ingestion, it has no complexity of pipeline orchestration and has a smother write workload.
+Most of the raw data set we are handling today are generating all the time in streaming way.
+
+Based on that, many requests for multiple writers' ingestion are derived. With multi-writer ingestion, several streaming events with the same schema can be drained into one Hudi table,
+the Hudi table kind of becomes a UNION table view for all the input data set. This is a very common use case because in reality, the data sets are usually scattered all over the data sources.
+
+Another very useful use case we wanna unlock is the real-time data set join. One of the biggest pain point in streaming computation is the dataset join,
+the engine like Flink has basic supports for all kind of SQL JOINs, but it stores the input records within its inner state-backend which is a huge cost for pure data join with no additional computations.
+In [HUDI-3304](https://issues.apache.org/jira/browse/HUDI-3304), we introduced a `PartialUpdateAvroPayload`, in combination with the lockless multi-writer,
+we can implement N-ways data sources join in real-time! Hudi would take care of the payload join during compaction service procedure.
+
+## Design
+
+### The Precondition
+
+#### MOR Table Type Is Required
+
+The table type must be `MERGE_ON_READ`, so that we can defer the conflict resolution to the compaction phase. The compaction service would resolve the conflicts of the same keys by respecting the event time sequence of the events.
+
+#### Deterministic Bucketing Strategy
+
+Determistic bucketing strategy is required, because the same records keys from different writers are desired to be distributed into the same bucket, not only for UPSERTs, but also for all the new INSERTs.
+
+#### Lazy Cleaning Strategy
+
+Config the cleaning strategy as lazy so that the pending instants are not rolled back by the other active writers.
+
+### Basic Work Flow
+
+#### Writing Log Files Separately In Sequence
+
+Basically, each writer flushes the log files in sequence, the log file rolls over for different versioning number,
+a pivotal thing needs to note here is that we need to make the write_token unique for the same version log files with the same base instant time,
+so that the file name does not conflict for the writers.
+
+The log files generated by a single writer can still preserve the sequence by versioning number, which is important if the natual order is needed for single writer events.
+
+![multi-writer](multi_writer.png)
+
+#### The Compaction Procedure
```

Comment:
as I know, for now Hudi does not support generate compaction plan during data writing, need to solve this problem

Reply:
> does not support generate compaction plan during data writing

Yes, we can only schedule a fresh new compaction plan after a commit/delta_commit had succeed, for `during data writing` did you indicate the procedure of on-going data writing?

Reply:
May be we can generate an alignment mark before  generate compaction plan, and the writer can scroll through the new file after discovering the mark. And then generate a schedule plan.

Reply:
> generate

@danny0405 yes, data should writing into new instant, hence the Instant generated by compaction plan in this situation, data will loss as long as some data writing to old instant after Compaction plan was generated.

@LinMingQiang  do you mean that mark will stop all the writers before a compaction plan ？

Reply:
The mark just tells writers to  write into a new instant file. If you have time, we can discuss this topic online.

Reply:
@danny0405  In the current single-writer mode, the compaction generation and the writer are actually serialized. If we want to support multiple writers, then we have to resolve the write conflicts between compaction generation and writers. Has this RFC considered this issue?